### ISSUE TITLE
Cannot set azurerm_key_vault name using expression / interpolated string

### ISSUE THREAD

**Issue by tkent** [hashicorp/terraform-provider-azurerm#17493]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

v1.0.5

### AzureRM Provider Version

3.12.0

### Affected Resource(s)/Data Source(s)

azurerm_key_vault

### Terraform Configuration Files

```hcl
variable "resource_group_name" {
  type        = string
  description = "The associated resource group id for module resources"
}

variable "env_key" {
  type        = string
  description = "The environment key"
}

terraform {
  required_providers {
    azurerm = {
      version = ">=3.12.0, < 4.0.0"
    }
    random = {
      version = ">=2.2.0"
    }
  }
}

resource "random_string" "deploy_token" {
  length  = 8
  special = false
  upper   = false
  lower   = true
  numeric = true
}

data "azurerm_resource_group" "rg" {
  name = var.resource_group_name
}

data "azurerm_client_config" "current" {}

locals {
  location     =  data.azurerm_resource_group.rg.location
  deploy_token = random_string.deploy_token.result
}

resource "azurerm_key_vault" "kv" {
  name                        = "${var.env_key}${local.deploy_token}"
  location                    = local.location
  resource_group_name         = data.azurerm_resource_group.rg.name
  enabled_for_disk_encryption = true
  tenant_id                   = data.azurerm_client_config.current.tenant_id
  soft_delete_retention_days  = 7
  purge_protection_enabled    = false
  sku_name                    = "standard"
}
```

### Debug Output/Panic Output

```shell
│ Error: "name" may only contain alphanumeric characters and dashes and must be between 3-24 chars
│ 
│   with azurerm_key_vault.kv,
│   on main.tf line 36, in resource "azurerm_key_vault" "kv":
│   36:   name                        = "${var.env_key}${local.deploy_token}"
│ 
╵
ERRO[0019] 1 error occurred:
	* exit status 1
```

### Expected Behaviour

The expression that defines the value for `name` should be evaluated before  validation rule is applied. 

### Actual Behaviour

The literal string `${var.env_key}${local.deploy_token}` is being tested against the validation rules. Manually setting the value to what the expression evaluates to works as expected.

### Steps to Reproduce

`terraform plan`

### Important Factoids

I bumped into this issue trying to upgrade a modules + resources defined in late 2021. When they were created, this expression evaluated correctly.

### References

I doubt it's relevant, but the Azure CLI versions on the systems we're using are `2.35.0` and `2.37.0`

**Comment by lonegunmanb** [comment]
Hi @tkent , may I ask what's the value for your `var.env_key` so I can reproduce this issue on my site? Thanks.

**Comment by lonegunmanb** [comment]
Hi @tkent , I've tried to reproduce this issue with following procedure get the same error:

1. Create a resource group named `zjhe-17493`
2. Create the following code file:

```hcl
provider "azurerm" {
  features {}
}

variable "resource_group_name" {
  type        = string
  description = "The associated resource group id for module resources"
}

variable "env_key" {
  type        = string
  description = "The environment key"
}

terraform {
  required_providers {
    azurerm = {
      version = ">=3.12.0, < 4.0.0"
    }
    random = {
      version = ">=2.2.0"
    }
  }
}

resource "random_string" "deploy_token" {
  length  = 8
  special = false
  upper   = false
  lower   = true
  numeric = true
}

data "azurerm_resource_group" "rg" {
  name = var.resource_group_name
}

data "azurerm_client_config" "current" {}

locals {
  location     =  data.azurerm_resource_group.rg.location
  deploy_token = random_string.deploy_token.result
}

resource "azurerm_key_vault" "kv" {
  name                        = "${var.env_key}${local.deploy_token}"
  location                    = local.location
  resource_group_name         = data.azurerm_resource_group.rg.name
  enabled_for_disk_encryption = true
  tenant_id                   = data.azurerm_client_config.current.tenant_id
  soft_delete_retention_days  = 7
  purge_protection_enabled    = false
  sku_name                    = "standard"
}
```

3. Execute `terraform plan -var env_key="thisisalongenv" -var resource_group_name="zjhe-17493"`:

```text
data.azurerm_client_config.current: Reading...
data.azurerm_resource_group.rg: Reading...
data.azurerm_client_config.current: Read complete after 0s [id=2022-07-06 01:51:27.0911878 +0000 UTC]
data.azurerm_resource_group.rg: Read complete after 1s [id=/subscriptions/85b3dbca-5974-4067-9669-67a141095a76/resourceGroups/zjhe-17493]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # azurerm_key_vault.kv will be created
  + resource "azurerm_key_vault" "kv" {
      + access_policy               = (known after apply)
      + enabled_for_disk_encryption = true
      + id                          = (known after apply)
      + location                    = "eastus"
      + name                        = (known after apply)
      + purge_protection_enabled    = false
      + resource_group_name         = "zjhe-17493"
      + sku_name                    = "standard"
      + soft_delete_retention_days  = 7
      + tenant_id                   = "72f988bf-86f1-41af-91ab-2d7cd011db47"
      + vault_uri                   = (known after apply)

      + network_acls {
          + bypass                     = (known after apply)
          + default_action             = (known after apply)
          + ip_rules                   = (known after apply)
          + virtual_network_subnet_ids = (known after apply)
        }
    }

  # random_string.deploy_token will be created
  + resource "random_string" "deploy_token" {
      + id          = (known after apply)
      + length      = 8
      + lower       = true
      + min_lower   = 0
      + min_numeric = 0
      + min_special = 0
      + min_upper   = 0
      + number      = true
      + numeric     = true
      + result      = (known after apply)
      + special     = false
      + upper       = false
    }

Plan: 2 to add, 0 to change, 0 to destroy.

─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── 

```

I think maybe you'll need to check your `var.env_key` value.

**Comment by tkent** [comment]
:wave: @lonegunmanb  

To your first question..

> Hi @tkent , may I ask what's the value for your var.env_key so I can reproduce this issue on my site? Thanks.

The value for the `env_key ` variable is `stage-cluster` in one instance. The entire resource name for that instance is "stage-cluster0waaduh2", but this was happening across multiple different keyvaults that were upgraded using the same module.

I should clarify that I used this module to create many key vaults about 9 months ago. Then, after upgrading the terraform provider + version and re-applying, I received the error I filed.

Since you're not seeing it, I decided to delete my statefile entirely and start over, re-importing the existing resources. That resolved the issue in the first one and I'm going to repeat it for all others. Since random statefile issues are not something anybody is up for looking into, I'm going to close this issue.

I'll close the case, since I don't think anybody wants to try to chase down a random statefile migration issue.
